### ISSUE TITLE
Bind API version with omniauth-facebook version

### ISSUE THREAD

**Issue by kenn** [mkdynamic/omniauth-facebook#190]
In README, it is suggested that we can specify an API version like `v2.2`, but it's not clear what is the best way actually.

```ruby
client_options: {
  site: 'https://graph.facebook.com/v2.2',
  authorize_url: 'https://www.facebook.com/v2.2/dialog/oauth' }
```

What I want to do is use the latest Facebook API version that is **tested with this gem's version**, not using unversioned API endpoints as it can break in the future.

So, say, if `omniauth-facebook-2.0.1` is tested with Facebook API v2.2, I want to use `v2.2`. Is it possible?

```ruby
facebook_api_version: :tested    # equivalent to 'v2.2' in this gem version
```

**Comment by simi** [comment]
Actually `omniauth-facebook` is not tested against real API endpoint via automated tests.

**Comment by kenn** [comment]
Here's what happened to me: I had been using the default setting, which is the unversioned API endpoints.

At a later time, I started to receive an exception that I hadn't seen, and the error was "unique constraint error for a user_id with multiple Facebook User IDs associated with the same user".

Turns out, Facebook updated the API at some point and now they return `app_scoped_user_id` instead of a real user ID, so our database now contains mixed data.

We couldn't fix this issue until we change the database schema and rewrite the data model.

By nature, unversioned API endpoints will bring in incompatible changes in the future, and I believe libraries like `omniauth-facebook` should offer a better default than this.

So, even though it's not really tested with a real API endpoint, I suggest that we have a required (not optional) config of the API version in string like `facebook_api_version: 'v2.2'`. We could even make the unversioned optionally available as `facebook_api_version: :unversioned`, but the point is, it's a choice that you consciously make. That way, the responsibility is delegated to the user of this library, and it won't break on production at a random time.

How does that sound?

**Comment by simi** [comment]
I think current behaviour is standard across omniauth-oauth(2) based gems. Also it is written in README.

> OmniAuth Facebook uses unversioned API endpoints by default. You can configure custom endpoints via client_options hash passed to provider.

We can add some shortcut to select API version, but I think it will be good to keep unversioned API as default one.

**Comment by kenn** [comment]
> I think current behaviour is standard across omniauth-oauth(2) based gems.

I don't think that's true, for instance, [omniauth-twitter hardcodes v1.1 path](https://github.com/arunagw/omniauth-twitter/blob/v1.1.0/lib/omniauth/strategies/twitter.rb#L34) and [omniauth-github hardcodes v3](https://github.com/intridea/omniauth-github/blob/v1.1.2/lib/omniauth/strategies/github.rb#L63).

> Also it is written in README.

It's my fault for sure that I missed it in README, but I learned it the hard way, by corrupting the integrity of the database. Considering how disastrous the consequence was, I don't see any reason that we'd ever want to use the unversioned API on production. The risk outweighs any potential benefits IMO.

**Comment by simi** [comment]
OK, are you able to provide PR for this?

**Comment by semaperepelitsa** [comment]
I don’t think it is a responsibility of an application to decide the API version at all. The library is making API calls, not my app. I guess, the problem is that there are no automated tests with the real API.

**Comment by simi** [comment]
let's decide what we need to close this

1. test against real API
2. lock to tested API

:question:

**Comment by dixalex** [comment]
Testing against real API would be better, but if no time for this - locking to tested API is also an option. The issue just needs to be resolved as soon as possible.

**Comment by mkdynamic** [comment]
Next version will use versioned APIs.

**Comment by kenn** [comment]
Awesome, thanks Mark! 👍

**Comment by svoop** [comment]
@mkdynamic Facebook keeps nagging due to the Aug 8 deadline for unversioned API access. What would you say, better switch to RC of version 4 now or is version 4 about to be released within the next couple of days anyways?

**Comment by mkdynamic** [comment]
@svoop Just released 4.0.0, since no bug reports from the RC. Let me know how it goes.

**Comment by svoop** [comment]
Cleaning my inbox, I come across this old issue. @mkdynamic Yep, worked like a charm, thanks a bunch for the fix!
